### DISCUSSION TITLE
Rar, 7zip and gz

### DISCUSSION THREAD

**Discussion by Yaume2** [PowerShell/PowerShell#23567] [Enhancements]
Hello, is Rar support planned for Expand-Archive?

**Comment by jdhitsolutions** [comment]
I would imagine part of that answer depends on if the API is open source. Although I thought I heard that Windows 11 was going to get native RAR support. So if that is the case, I would think PowerShell would be able to leverage those APIs.

**Comment by SteveL-MSFT** [comment]
There are modules on https://powershellgallery.com that support those formats

**Comment by kasini3000** [comment]
Welcome to my Powershell hybrid cloud batch management project: kasini3000
Welcome to start me
It integrates 7zip for both win and Linux. It will copy the "node_script" directory to each controlled computer (win Linux). 
It can be used independently.   
https://gitee.com/chuanjiao10/kasini3000/tree/master/node_script/7z
For RAR decompression:
L7zx. ps1
W7zx.ps1

**Comment by AdreKiseque** [comment]
24H2 has native support for rar and 7zip now, it would be good for those features to be added natively to PowerShell now as well I think

**Reply by rhubarb-geek-nz** [reply]
I would like to move away from the idea that the core PowerShell has to support everything. It is supposed to be a shell, not a monolithic toolkit containing everything anyone might possibly need.

**Reply by AdreKiseque** [reply]
Is it a crazy idea that a shell should be able to do the same things that can be done natively through the GUI?

**Reply by rhubarb-geek-nz** [reply]
> Is it a crazy idea that a shell should be able to do the same things that can be done natively through the GUI?

It's crazy that a shell is expected to duplicate every command line program. Rar and 7zip are available as command line programs that can already be invoked directly from PowerShell. It is crazy that a shell is supposed to re-implement what is already available in PSGallery. PowerShell 7 is not part of the Windows operating system, it is a cross platform shell.

**Reply by AdreKiseque** [reply]
Right, but why should you need to install a separate program for that anyway? Handling archives is a pretty basic functionality, and one that can be done natively on the OS. I believe you should be able to do anything that can be done natively through the GUI natively through the command line.

**Reply by rhubarb-geek-nz** [reply]
> Right, but why should you need to install a separate program for that anyway?

You are asking for the functionality to be added to a program (pwsh.exe) that does not come with Windows and you have to install separately.

> I believe you should be able to do anything that can be done natively through the GUI natively through the command line.

You are welcome to believe that, but that is not how Windows is built. On Linux GUI tools are typically wrappers on command line tools. This is not the case in Windows.

If you like, that is how the original PowerShell came into being, because Windows had to be managed through a GUI because Windows does not provide command line tools for everything.

Even on Linux, where what you describe is the case, each application is literally downloaded separately from apt or yum repositories. Bash or zsh do not have to be continually extended, it is the applications themselves that provide the functionality. In the PowerShell world the parallel to this is downloading from a NuGet repository such as PSGallery.

**Reply by AdreKiseque** [reply]
> If you like, that is how the original PowerShell came into being, because Windows had to be managed through a GUI because Windows does not provide command line tools for everything.

Well then, this change seems in-line with PowerShell's original goal then, no?

> You are asking for the functionality to be added to a program (pwsh.exe) that does not come with Windows and you have to install separately.

Fair point, though I argue PowerShell 7 should ship in-box with Windows as well.

**Reply by rhubarb-geek-nz** [reply]
> Well then, this change seems in-line with PowerShell's original goal then, no?

PowerShell would collapse under its own weight if it had to wrap or duplicate everything that Windows provided. When PowerShell Core was split from PowerShell Desktop much of the Windows specific code was stripped out to become a cross platform shell.

PowerShell is extensible, it can be extended by scripts or modules, or by calling existing executables. It does not have to contain everything to start with.

PowerShell is built on dotnet, if things are not made easily available in the dotnet SDK then it becomes much harder to provide. Windows itself is not built using dotnet, Windows operating system and libraries are still effectively C/C++/COM. 

[Expand-Archive](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.archive/expand-archive?view=powershell-7.4) is provided by the frozen repository [Microsoft.PowerShell.Archive](https://github.com/PowerShell/Microsoft.PowerShell.Archive), so I suggest to look to the gallery for support for other formats or workarounds.

Literally the only archive format that PowerShell needs to support is ZIP so that it can download and install NuGet packages. Once you have that you have extensibility, but it is still provided by a module as is the cmdlet to install a module.

**Comment by kasini3000** [comment]
in my submission:
1 It's time to integrate command line decompression with 7z and rar format libraries in Win11.
It depends on how Microsoft does it. Should MS donate to the zip project or to other projects?

2 After obtaining the command-line library, the remaining task is to update Expand Archive to support it.

3 Microsoft and Expand Archive seem to be indifferent to the Linux version of PowerShell how to decompress rar and 7z.
so： 
For Windows, Linux,  uniformity, batch automatic operation and maintenance. 
Welcome to use the kasini3000 framework I have written.
